File: main/classes/core/src/com/ibm/icu/util/TimePeriod.java
/*
 *******************************************************************************
 * Copyright (C) 2013, International Business Machines Corporation and         *
 * others. All Rights Reserved.                                                *
 *******************************************************************************
 */
package com.ibm.icu.util;

import java.util.Arrays;
import java.util.Iterator;
import java.util.NoSuchElementException;

/**
 * TimePeriod represents a time period.  TimePeriod objects are immutable.
 * <p>Example usage:
 * <pre>
 *   Period p = Period.forAmounts(
 *       new TimeUnitAmount (TimeUnit.WEEK, 5),
 *       new TimeUnitAmount (TimeUnit.DAY, 40),
 *       new TimeUnitAmount (TimeUnit.HOUR, 2),
 *       new TimeUnitAmount (TimeUnit.SECOND, 8));
 * </pre>
 * @draft ICU 52
 * @provisional This API might change or be removed in a future release.
 */
public final class TimePeriod implements Iterable<TimeUnitAmount> {
    
    private final TimeUnitAmount[] fields;
    private final int size;
    private final int hash;
    
    private TimePeriod(TimeUnitAmount[] fields, int size, int hash) {
        this.fields = fields;
        this.size = size;
        this.hash = hash;
    }
       
    /**
     * Returns a new TimePeriod that matches the given time unit amounts.
     * @param amounts the TimeUnitAmounts. Must be non-empty. Normalization of the
     *   amounts and inclusion/exclusion of 0 amounts is up to caller. The Number
     *   in each TimeUnitAmount must either be a Byte, Short, Integer, Long, Float,
     *   Double, BigInteger, or BigDecimal or it must implement Cloneable and have
     *   a public clone method.
     * @return the new TimePeriod object
     * @throws IllegalArgumentException if multiple TimeUnitAmount objects match
     * the same time unit or if any but the smallest TimeUnit has a fractional value
     * Or if amounts is empty.
     * @draft ICU 52
     * @provisional This API might change or be removed in a future release.
     */
    public static TimePeriod forAmounts(TimeUnitAmount ...amounts) {
        return forAmounts(Arrays.asList(amounts));      
    }

    /**
     * Returns a new TimePeriod that matches the given time unit amounts.
     * @param amounts the TimeUnitAmounts. Must be non-empty. Normalization of the
     *   amounts and inclusion/exclusion of 0 amounts is up to caller. The Number
     *   object in each TimeUnitAmount must not change. Otherwise the created
     *   TimePeriod object may not work as expected.
     * @return the new TimePeriod object
     * @throws IllegalArgumentException if multiple TimeUnitAmount objects match
     * the same time unit or if any but the smallest TimeUnit has a fractional value
     * Or if amounts is empty.
     * @draft ICU 52
     * @provisional This API might change or be removed in a future release.
     */
    public static TimePeriod forAmounts(Iterable<TimeUnitAmount> amounts) {
        TimeUnitAmount[] fields = new TimeUnitAmount[TimeUnit.TIME_UNIT_COUNT];
        int size = 0;
        for (TimeUnitAmount tua : amounts) {
            int index = tua.getTimeUnit().getIndex();
            if (fields[index] != null) {
                throw new IllegalArgumentException(
                        "Only one TimeUnitAmount per unit allowed.");
            }
            // This line is necessary to guarantee immutability of the TimePeriod
            // class. A Number object, which is in TimeUnitAmount, need not be immutable,
            // but Double is immutable.
            fields[index] = tua;
            size++;
        }
        if (size == 0) {
            throw new IllegalArgumentException(
                    "There must be at least one TimeUnitAmount.");
        }
        TimePeriod result = new TimePeriod(fields, size, computeHash(fields));
        boolean fractionalFieldEncountered = false;
        for (TimeUnitAmount tua : result) {
            if (fractionalFieldEncountered) {
                throw new IllegalArgumentException(
                    "Only the smallest time unit can have a fractional amount.");
            }
            double value = tua.getNumber().doubleValue();
            if (value != Math.floor(value)) {
                fractionalFieldEncountered = true;
            }
        }
        return result;
    }
        
    /**
     * Gets the value for a specific time unit.
     * @param timeUnit the time unit.
     * @return the TimeUnitAmount or null if no value is present for given TimeUnit.
     *  A non-existing value and a zero value are two different things.
     * @draft ICU 52
     * @provisional This API might change or be removed in a future release.
     */
    public TimeUnitAmount getAmount(TimeUnit timeUnit) {
        return fields[timeUnit.getIndex()];
    }

    /**
     * Returned iterator iterates over all TimeUnitAmount objects in this object.
     * Iterated TimeUnitAmount objects are ordered from largest TimeUnit to
     * smallest TimeUnit. Remove method on returned iterator throws an
     * UnsupportedOperationException.
     * @draft ICU 52
     * @provisional This API might change or be removed in a future release.
     */
    public Iterator<TimeUnitAmount> iterator() {
      return new TPIterator();
    }

    /**
     * Returns the number of TimeUnitAmount objects in this object.
     * @internal
     * @deprecated This API is ICU internal only.
     */
    public int size() {
      return size;
    }
    
    /**
     * Two TimePeriod objects are equal if they contain equal TimeUnitAmount objects.
     * @draft ICU 52
     * @provisional This API might change or be removed in a future release.
     */
    @Override
    public boolean equals(Object o) {
        if (this == o) {
            return true;
        }
        if (!(o instanceof TimePeriod)) {
            return false;
        }
        TimePeriod rhs = (TimePeriod) o;
        if (this.hash != rhs.hash) {
            return false;
        }
        return Arrays.equals(fields, rhs.fields);
    }

    /**
     * Overrides hashCode().
     * @draft ICU 52
     * @provisional This API might change or be removed in a future release.
     */
    @Override
    public int hashCode() {
        return hash;
    }
    
    private static int computeHash(TimeUnitAmount[] fields) {
        int result = 0;
        for (TimeUnitAmount amount : fields) {
            result *= 31;
            if (amount != null) {
                result += amount.hashCode();
            }
        }
        return result;
    }
    
    private class TPIterator implements Iterator<TimeUnitAmount> {
        
        private int index = 0;

        public boolean hasNext() {
            while (index < fields.length && fields[index] == null) {
                index++;
            }
            return index < fields.length;
        }

        public TimeUnitAmount next() {
            if (!hasNext()) {
                throw new NoSuchElementException();
            }
            return fields[index++];
        }

        public void remove() {
            throw new UnsupportedOperationException();           
        }
    } 
}

File: source/i18n/timeperiod.cpp
/*
 *******************************************************************************
 * Copyright (C) 2013-2013, Google, International Business Machines Corporation
 * and others. All Rights Reserved.
 *******************************************************************************
 */

#include <math.h>
#include "unicode/timeperiod.h"
#include "putilimp.h"

#if !UCONFIG_NO_FORMATTING

U_NAMESPACE_BEGIN

TimePeriod::TimePeriod(const TimePeriod& other) : fLength(other.fLength) {
    for (int32_t i = 0; i < sizeof(fFields) / sizeof(fFields[0]); ++i) {
        if (other.fFields[i] == NULL) {
            fFields[i] = NULL;
        } else {
            fFields[i] = (TimeUnitAmount *) other.fFields[i]->clone();
        }
    }
}


TimePeriod::TimePeriod(
        const TimeUnitAmount * const *timeUnitAmounts,
        int32_t length,
        UErrorCode& status) : fLength(0) {
    int32_t fieldCount = sizeof(fFields) / sizeof(fFields[0]);
    for (int32_t i = 0; i < fieldCount; ++i) {
        fFields[i] = NULL;
    }
    if (U_FAILURE(status)) {
        return;
    }
    for (int32_t i = 0; i < length; ++i) {
        const TimeUnitAmount *tua = timeUnitAmounts[i];
        int32_t idx = tua->getTimeUnitField();
        if (fFields[idx] != NULL) {
            status = U_ILLEGAL_ARGUMENT_ERROR;
            return;
        }
        fFields[idx] = (TimeUnitAmount *) tua->clone();
        ++fLength;
    }
    validate(status);
    if (U_FAILURE(status)) {
        return;
    }
}

void TimePeriod::validate(UErrorCode& status) const {
    if (U_FAILURE(status)) {
        return;
    }
    if (fLength == 0) {
        status = U_ILLEGAL_ARGUMENT_ERROR;
        return;
    }  
    int32_t fieldCount = sizeof(fFields) / sizeof(fFields[0]);
    UBool fractionalFieldEncountered = FALSE;
    for (int32_t i = 0; i < fieldCount; ++i) {
        if (fFields[i] == NULL) {
             continue;
        }
        if (fractionalFieldEncountered) {
            status = U_ILLEGAL_ARGUMENT_ERROR;
            return;
        }
        UErrorCode doubleStatus = U_ZERO_ERROR;
        double value = fFields[i]->getNumber().getDouble(doubleStatus);
        if (value != uprv_floor(value)) {
            fractionalFieldEncountered = TRUE;
        }
    }
}

TimePeriod::~TimePeriod() {
    int32_t fieldCount = sizeof(fFields) / sizeof(fFields[0]);
    for (int32_t i = 0; i < fieldCount; ++i) {
        delete fFields[i];
    }
}

UBool TimePeriod::operator==(const TimePeriod& other) const {
    // If same object, they are equal
    if (this == &other) {
        return TRUE;
    }
    int32_t fieldCount = sizeof(fFields) / sizeof(fFields[0]);
    for (int32_t i = 0; i < fieldCount; ++i) {
        if (fFields[i] == other.fFields[i]) {
            continue;
        }    
        // One is NULL, the other isn't.
        if (fFields[i] == NULL || other.fFields[i] == NULL) {
            return FALSE;
        }
        if (*fFields[i] != *other.fFields[i]) {
            return FALSE;
        }
    }
    return TRUE;
}
    
UBool TimePeriod::operator!=(const TimePeriod& other) const {
    return !(*this == other);
}

const TimeUnitAmount *TimePeriod::getAmount(
        TimeUnit::UTimeUnitFields field) const {
    return fFields[field];
}

U_NAMESPACE_END

#endif
